File: face/geomajas-face-dojo/client/src/main/resources/geomajas/action/menu/editing/InsertPointAction.js
dojo.provide("geomajas.action.menu.editing.InsertPointAction");
/*
 * This is part of Geomajas, a GIS framework, http://www.geomajas.org/.
 *
 * Copyright 2008-2011 Geosparc nv, http://www.geosparc.com/, Belgium.
 *
 * The program is available in open source according to the GNU Affero
 * General Public License. All contributions in this program are covered
 * by the Geomajas Contributors License Agreement. For full licensing
 * details, see LICENSE.txt in the project root.
 */
dojo.require("geomajas.action.Action");

dojo.declare("InsertPointAction", Action, {

	/**
	 * @fileoverview Insert a point from a feature's geometry (rightmouse menu).
	 * @class Insert a point in a feature's geometry.
	 * @author Pieter De Graef
	 *
	 * @constructor 
	 * @extends Action
	 * @param id This action's unique identifier.
	 * @param mapWidget Reference to a MapWidget object.
	 * @param index Every point in a geometry has an index.
	 * @param position Coordinate holding the position for the new point.
	 */
	constructor : function (id, mapWidget, index, position) {
		/** @private */
		this.mapWidget = mapWidget;

		/** @private */
		this.index = index;

		/** Unique identifier */
		this.id = id;

		/** The action can be displayed as text only. */
		this.text = "Insert point";

		/** The point's position. */
		this.position = position;
	},

	/**
	 * Execute a {@link InsertCoordinateCommand} on the feature in the
	 * {@link FeatureTransaction}, stored in the {@link FeatureEditor}. Also
	 * redraws the {@link FeatureTransaction}.
	 * @param event The {@link HtmlMouseEvent} from clicking this action.
	 */
	actionPerformed : function (event) {
		var trans = this.mapWidget.getMapModel().getFeatureEditor().getFeatureTransaction();
		if (trans != null) {
			dojo.publish (this.mapWidget.getRenderTopic(), [trans, "delete"]);
			var command = new InsertCoordinateCommand (this.index, this.position);
			trans.execute (command);
			dojo.publish (this.mapWidget.getRenderTopic(), [trans, "all"]);
		}
	}
});

File: client/geomajas-client-smartgwt/client/src/main/java/org/geomajas/smartgwt/client/action/menu/InsertPointAction.java
/*
 * This is part of Geomajas, a GIS framework, http://www.geomajas.org/.
 *
 * Copyright 2008-2013 Geosparc nv, http://www.geosparc.com/, Belgium.
 *
 * The program is available in open source according to the GNU Affero
 * General Public License. All contributions in this program are covered
 * by the Geomajas Contributors License Agreement. For full licensing
 * details, see LICENSE.txt in the project root.
 */

package org.geomajas.smartgwt.client.action.menu;

import org.geomajas.geometry.Coordinate;
import org.geomajas.smartgwt.client.action.MenuAction;
import org.geomajas.smartgwt.client.gfx.MenuContext;
import org.geomajas.smartgwt.client.i18n.I18nProvider;
import org.geomajas.smartgwt.client.map.feature.FeatureTransaction;
import org.geomajas.smartgwt.client.map.feature.TransactionGeomIndex;
import org.geomajas.smartgwt.client.map.feature.TransactionGeomIndexUtil;
import org.geomajas.smartgwt.client.map.feature.operation.InsertCoordinateOp;
import org.geomajas.smartgwt.client.util.WidgetLayout;
import org.geomajas.smartgwt.client.widget.MapWidget;

import com.smartgwt.client.widgets.Canvas;
import com.smartgwt.client.widgets.menu.Menu;
import com.smartgwt.client.widgets.menu.MenuItem;
import com.smartgwt.client.widgets.menu.MenuItemIfFunction;
import com.smartgwt.client.widgets.menu.events.MenuItemClickEvent;

/**
 * Insert a new point in the geometry at a given index. This action will take the target from the context menu event,
 * and create an index from it.
 * 
 * @author Pieter De Graef
 */
public class InsertPointAction extends MenuAction implements MenuItemIfFunction {

	private MapWidget mapWidget;

	private TransactionGeomIndex index;

	private Coordinate coordinate;

	/**
	 * Create the menu action for inserting points in a geometry.
	 * 
	 * @param mapWidget
	 *            The map on which editing is going on.
	 */
	public InsertPointAction(MapWidget mapWidget) {
		super(I18nProvider.getMenu().insertPoint(), WidgetLayout.iconVertexCreate);
		this.mapWidget = mapWidget;
		setEnableIfCondition(this);
	}

	/**
	 * Insert a new point in the geometry at a given index. This action will take the target from the context menu
	 * event, and create an index from it. Then insert a new coordinate with the position equal to the context menu
	 * event's position at the index.
	 * 
	 * @param event
	 *            The {@link MenuItemClickEvent} from clicking the action.
	 */
	public void onClick(MenuItemClickEvent event) {
		FeatureTransaction ft = mapWidget.getMapModel().getFeatureEditor().getFeatureTransaction();
		if (ft != null && index != null && coordinate != null) {
			mapWidget.render(ft, MapWidget.RenderGroup.VECTOR, MapWidget.RenderStatus.DELETE);
			InsertCoordinateOp op = new InsertCoordinateOp(index, coordinate);
			ft.execute(op);
			mapWidget.render(ft, MapWidget.RenderGroup.VECTOR, MapWidget.RenderStatus.ALL);
		}
	}

	/**
	 * Implementation of the <code>MenuItemIfFunction</code> interface. This will determine if the menu action should be
	 * enabled or not. In essence, this action will be enabled when the context menu event occurred on an edge of the
	 * painted <code>FeatureTransaction</code>.
	 */
	public boolean execute(Canvas target, Menu menu, MenuItem item) {
		FeatureTransaction featureTransaction = mapWidget.getMapModel().getFeatureEditor().getFeatureTransaction();
		if (featureTransaction != null) {
			MenuContext graphics = mapWidget.getMenuContext();
			coordinate = mapWidget.getMapModel().getMapView().getWorldViewTransformer().viewToWorld(
					graphics.getRightButtonCoordinate());
			String targetId = graphics.getRightButtonName();
			if (targetId != null && TransactionGeomIndexUtil.isEdge(targetId)) {
				index = TransactionGeomIndexUtil.getIndex(targetId);
				return true;
			}
		}
		return false;
	}
}
